rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,289 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: draft
+---
+
+# RFD 57 - Automatic discovery and enrollment of AWS servers
+
+## What
+
+Proposes a way by which an SSH service might automatically discover and register AWS
+EC2 instances.
+
+## Why
+
+Currently when adding a new AWS server, it's required that Teleport be installed
+after the server has been provisioned which may be a slow process for organizations
+with large numbers of servers as it needs to be installed and then added to the
+teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve the
+issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+A Teleport SSH agent will need to be configured with an assume role policy for each
+of the accounts that the agent will discover services on as described by the IAM join
+method docs[1]
+
+Discovery can use a matcher similar to the `db_service/aws` matcher, however EC2
+instances will have an optional install command:
+
+```yaml
+ssh_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes" # aws tags to match
+    ssm_command_document: ssm_command_document_name
+    accounts:
+    - aws_account: "222222222222"
+      aws_roles:
+      - "arn:aws:iam::222222222222:role/teleport-DescribeInstances-role"
+      - "arn:aws:iam::222222222222:role/teleport-Install-role"
+```
+
+The agent will use EC2's `DescribeInstances` API in order to list instances[1]. This
+will require the teleport SSH agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions
+
+As with AWS database discover, new EC2 nodes will be discovered periodically on a 60
+second timer, as new nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall teleport on top of an instance where it is
+already present the generated teleport config will include a static label indicating
+it was created via auto discovery.
+
+Example:
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "AWS_INSTANCE_ID",
+    "labels": {
+      "env": "example"
+      "teleport.dev/discovered-node": "yes"
```

Comment:
We have a `teleport.dev/origin` label that discovered databases set to `cloud`. Can we reuse it here as well so we don't introduce another label?